File: front-end/src/app/domain/models/game/action/impl/gamePlay/StartGameAction.ts
import {GameAction} from "../../GameAction";
import {Player} from "../../../Player";

/**
 * TODO: Add JsDoc
 *
 * @author Hamza el Haouti
 */
export class StartGameAction extends GameAction {
  private _startCapital: number;
  private _firstTurnEndsAt: Date;

  public constructor(
    id: number,
    actor: Player,
    performedAt: Date,
    startCapital: number,
    firstTurnEndsAt: Date
  ) {
    super(StartGameAction.name, id, actor, performedAt);
    this._startCapital = startCapital;
    this._firstTurnEndsAt = firstTurnEndsAt;
  }

  public static trueCopy(action: StartGameAction): GameAction | null {
    return new StartGameAction(
      action.id,
      Player.trueCopy(action.actor),
      new Date(action.performedAt),
      action.startCapital,
      new Date(action.firstTurnEndsAt)
    );
  }

  public override toJson(): object {
    return Object.assign(
      super.toJson(), {
        "startCapital": this.startCapital,
        "firstTurnEndsAt": this.firstTurnEndsAt.toJSON(),
      });
  }

  public get startCapital(): number {
    return this._startCapital;
  }


  public get firstTurnEndsAt(): Date {
    return this._firstTurnEndsAt;
  }

  public override toString(): string {
    return `${this.actor.user.username} started the game. The start capital is $${this.startCapital}`;
  }

}

File: back-end/src/main/java/nl/hva/backend/domain/models/game/action/impl/gamePlay/StartGameAction.java
package nl.hva.backend.domain.models.game.action.impl.gamePlay;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import nl.hva.backend.domain.models.game.Player;
import nl.hva.backend.domain.models.game.action.GameAction;

import javax.persistence.Entity;
import java.time.LocalDateTime;

/**
 * TODO: Add JavaDoc
 *
 * @author Hamza el Haouti
 */
@Entity
@Getter
@Setter
@NoArgsConstructor
public class StartGameAction extends GameAction {
    private double startCapital;

    private LocalDateTime firstTurnEndsAt;

    public StartGameAction(
            Player actor,
            double startCapital,
            LocalDateTime firstTurnEndsAt
    ) {
        super(actor);
        this.startCapital = startCapital;
        this.firstTurnEndsAt = firstTurnEndsAt;
    }

}